eventuate-tram-cdc-connector/src/main/java/io/eventuate/tram/cdc/connector/pipeline/CdcTramPipelineFactoryConfiguration.java

```diff
@@ -25,6 +26,6 @@ public CdcPipelineFactory cdcPipelineFactory(DataProducerFactory dataProducerFac
                     publishingFilter,
                     new MessageWithDestinationPublishingStrategy(),
                     meterRegistry),
-            new BinlogEntryToMessageConverter());
+            readerId -> new BinlogEntryToMessageConverter(new DatabaseIdGenerator(readerId)));
```

Comment:
I am confused by this.
readerId can be null, right?
DatabaseIdGenerator's constructor should/does verify that readerId is not null, right - since it needs one.

Reply:
Chris,

tests uses both id generators, application and database.
Application id generator is used when reader id is not specified, that is why here null is allowed:
https://github.com/eventuate-foundation/eventuate-cdc/pull/75/files#diff-0dfbeeb39750b2d0117def81e57cbe28R20
Also it is allowed for database binlog clients.
But the cdc service requires reader id: https://github.com/eventuate-foundation/eventuate-cdc/pull/75/files#diff-fafb9b7680f553fb1b9f65d6bf4202c8R30

Regarding that line of code (29), reader id is Long (not long) because 

BinlogEntryToEventConverterFactory (specified as lambda on line 29) is Function<Long, BinlogEntryToEventConverter<EVENT>>

and as we know Generics cannot be primitives.
Also I don't think that is critical to change BinlogEntryToEventConverterFactory to use explicit method definition.

Reply:
I think you are missing understanding.

`readerId -> new BinlogEntryToMessageConverter(new DatabaseIdGenerator(readerId))` creates a BinlogEntryToMessageConverter

But what if `readerId` is null - i.e. a reader for a schema using old id generation

Reply:
My logic was that when we migrate the cdc we should anyway specify a reader id.
But in case if messages were generated with application id, cdc does nothing, because ID header is present: 

https://github.com/eventuate-foundation/eventuate-cdc/blob/8a30146aedd246e34b33602c2fd88f2b3b880b59/eventuate-tram-cdc-connector/src/main/java/io/eventuate/tram/cdc/connector/BinlogEntryToMessageConverter.java#L24-L27

Reply:
So you have added a new mandatory property for a reader: `id` which is in reality `outboxId`.
Not sure that makes sense if someone is using application generated IDs and they haven't thought about migrating to db generated IDs.

Reply:
I think it makes migration more straightforward.
As I see a migration scenario:

1) User updates all libraries, including cdc.
2) Migrates schema.
3) Cdc forces him to specify unique id for reader (user should not think about it much, he should only know that id is unique and has 48 unsigned bits).
4) Everything works as before (application generated id)
5) At some point that user desides to use database generated id. He only uses outbox id for applications as he used for  reader. Because other work already done during an update.